### PR TITLE
DEV-2445 — Server-Side changes for DEV-1869

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This is one of two planned stacked PRs to accomplish the AC in the parent ticket, [DEV-1869
](https://news-revenue-hub.atlassian.net/jira/software/c/projects/DEV/boards/3?modal=detail&selectedIssue=DEV-1869)

This will be merged into DEV-1869-root-branch, and will also serve as the starting point for the next ticket, [DEV-2446](https://news-revenue-hub.atlassian.net/browse/DEV-2446), which is where the SPA side changes for DEV-1869 will be made.

#### What's this PR do?

- Adds `custom_thank_you_page_enabled` as a property of the `Plan` dataclass, defaulting to `False`
- Blocks Django admin from setting `thank_you_redirect` on a page whose parent Org is not provided this feature by its plan, and displays error message accordingly.
- Blocks Django admin from creating a page if parent org of referenced RP is at its plan-determined page limit. NB: This was not part of original AC, but it's something that was missed in DEV-2178, and was easy to do here.
- Adds org plan to page serializer so plan-related validation can be more easily done on the front end on a per page basis.
- Adds tests for admin- and serializer-related changes

#### Why are we doing this? How does it help us?

Part of rolling out self service onboarding.

#### How should this be manually tested? Please include detailed step-by-step instructions.

For page limit related validation:

1. Ensure that BillyPenn org is on the Free plan [here](https://dev-2445.revengine-review.org/nrhadmin/organizations/organization/2/change/).  
2. Try to add a new page for BillyPenn [here](https://dev-2445.revengine-review.org/nrhadmin/pages/donationpage/add/).
3. You should see an error message saying BillyPenn's plan does not allow for more pages.
4. Update BillyPenn org's plan to "PLUS" [here](https://dev-2445.revengine-review.org/nrhadmin/organizations/organization/2/change/). 
5. Try to add a new page for BillyPenn [here](https://dev-2445.revengine-review.org/nrhadmin/pages/donationpage/add/). This time it should succeed.

For thank you redirect related validation:

1. Ensure that BillyPenn org is on the Free plan [here](https://dev-2445.revengine-review.org/nrhadmin/organizations/organization/2/change/). 
2. Edit this [billypenn page](https://dev-2445.revengine-review.org/nrhadmin/pages/donationpage/1/change/), pasting in https://www.google.com (or some other valid URL) into the `thank_you_redirect` field. Click Save. You should see an error message indicating that this feature is not enabled by the plan.
3. Update BillyPenn org's plan to "PLUS" [here](https://dev-2445.revengine-review.org/nrhadmin/organizations/organization/2/change/). 
4. Do Step 2 again. This time it should succeed.

The serializer related changes can be verified by s[igning in to the SPA in review app here](https://dev-2445.revengine-review.org/sign-in) and then going (in same or separate tab)  [here](https://dev-2445.revengine-review.org/api/v1/pages/1/). You should see a serialized page, and note the presence of a nested `plan` property.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

N/A

#### What are the relevant tickets? Add a link to any relevant ones.

- [DEV-2445](https://news-revenue-hub.atlassian.net/browse/DEV-2445)
- [DEV-1869](https://news-revenue-hub.atlassian.net/browse/DEV-1869)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

Tracked in parent ticket DEV-1869.

### CODE REVIEW

File: apps/api/tests/__init__.py [hunk 1]
```diff
@@ -156,7 +156,7 @@ def assert_superuser_can_patch(self, url, data):
 
     def assert_superuser_cannot_patch(self, url, data, expected_status_code=status.HTTP_404_NOT_FOUND):
         data = data if data is not None else {}
-        return self.assert_user_can_patch(url, self.superuser, data)
+        return self.assert_user_cannot_patch(url, self.superuser, data, expected_status_code)
```
Comment (njharman) [inline apps/api/tests/__init__.py:159]: Use pytest assert style testing on new and modified code

File: apps/organizations/models.py [hunk 2]
```diff
@@ -39,6 +39,7 @@ class Plan:
     name: str
     label: str
     page_limit: int = 1
+    custom_thank_you_page_enabled: bool = False
```
Comment (njharman) [inline apps/organizations/models.py:42]: These Plan instances are "bags" on constants?  Not meant to change?

Consider https://docs.python.org/3/library/dataclasses.html#frozen-instances

**Review by njharman — CHANGES_REQUESTED** [review]

File: apps/api/tests/__init__.py (hunk 1 above)
Comment (benjaminEwhite) [reply apps/api/tests/__init__.py:159]: This is a convenience function that does a number of things "wrapping" the assertion itself. While the `assert_user_cannot_patch` should be refactored when/if touched, refactoring that function to use pytest assert is not in scope for this PR.

File: apps/organizations/models.py (hunk 2 above)
Comment (benjaminEwhite) [reply apps/organizations/models.py:42]: 53d9e312

**Review by x110dc — APPROVED** [review]

**Pull request merged**